### ISSUE TITLE
reactivity on meteor login

### ISSUE THREAD

**Issue by ujwal-setlur** [inProgress-team/react-native-meteor#195]
I am sometimes (~25% of the times) noticing that once my react-native app logs into meteor, it doesn't pick up on the fact that the login is successful and go to the next scene. I think it is a react-native-meteor issue, but I am not sure. One indication is that when I quit the app, and restart it, it auto-logs into meteor and sends me to the next scene as designed.

Is this the fact that reactivity on the Meteor login is failing?

Is there a way to debug this? Here is my code snippet:
```
class MainComponent extends Component {

  componentWillMount() {

    // Connect to Meteor backend!
    Meteor.connect(settings.METEOR_URL);

    // Configure Google SignIn
    GoogleSignin.hasPlayServices({
      autoResolve: true
    }).then(() => {
      // play services are available. can now configure library
      GoogleSignin.configure({
        iosClientId: settings.google.iosClientId,
        webClientId: settings.google.webClientId,
        offlineAccess: false
      });
    })
      .catch((err) => {
        console.log("Play services error", err.code, err.message);
      });

    // Set up our scenes
    this.scenes = Actions.create(
      <Scene
        key="root"
        component={createContainer(this.composer, Switch)}
        selector={this.selector}
        tabs={true}
        unmountScenes
      >
        <Scene
          key="connecting"
          hideNavBar={true}
          component={Routes.loading.component}
          title={Routes.loading.title}
        />
        <Scene key="loggedOut">
          <Scene
            key={Routes.signIn.key}
            component={Routes.signIn.component}
            title={Routes.signIn.title}
            hideNavBar={true}
            initial
          />
        </Scene>
        <Scene key="loggedIn">
          <Scene
            key={Routes.home.key}
            component={Routes.home.component}
            title={Routes.home.title}
            hideNavBar={true}
            initial
          />
        </Scene>
      </Scene>
    );
  }

  composer() {
    return {
      connected: Meteor.status().connected,
      user: Meteor.user()
    };
  }

  /*eslint-disable no-unused-vars*/
  selector(data, props) {
    if (!data.connected) {
      return "connecting";
    } else if (!data.user) {
      return "loggedOut";
    } else {
      return "loggedIn";
    }
  }
  /*eslint-enable no-unused-vars*/

  render() {
    return (
      <Router scenes={this.scenes} />
      );
  }
}

export default MainComponent;
```

**Comment by sabativi** [comment]
This is also a problem I am facing, i am guessing that when the app has been put in background for a certain amount of time, the ddp connection is closed.

When you open the app again, we wait for the user to be loggedin before displaying the loggedIn view...

To be honest I already have the problem on an ionic app that have been written with angular-meteor project, do not know how to solve this properly.

Now maybe using this project we can just check if the auth token see https://github.com/inProgress-team/react-native-meteor/pull/129 is set and use https://github.com/DesignmanIO/react-native-meteor-redux to get data even when the user if offline

What do you think about that ?

**Comment by JulianKingman** [comment]
It looks like you should be using state, instead of the `selector()` method. Here's what I do, which works quite well (similar to @sabativi 's suggestion): 
````jsx
componentWillMount() {
    // wait for login token
    AsyncStorage.getItem('reactnativemeteor_usertoken').then((loginToken) => {
      if (loginToken) {
        // if token exists, they're logged in, take them to the initial route
        this.setState({initialRoute: Router.getRoute('drawer').routeName});
        this.setState({loggedIn: true});
      } else {
       // if token doesn't exist, they're not logged in, so take them to sign in screen
        this.setState({initialRoute: Router.getRoute('signIn').routeName});
        this.setState({loggedIn: false});
      }
      // either way, the token is done loading
      this.setState({isLoading: false});
    }).done();
  }

  render() {
    if (this.state.isLoading) return <Loading />;
    return (
      <StackNavigation initialRoute={this.state.initialRoute} />
    )
````
I'm using ex-navigation, so you're router may look different.

**Comment by ujwal-setlur** [comment]
Thanks! I will try that. But was what I was doing wrong?

Get Outlook for iOS




On Fri, Jan 20, 2017 at 5:45 AM -0800, "JulianKingman" <notifications@github.com> wrote:












It looks like you should be using state, instead of the selector() method. Here's what I do, which works quite well (similar to @sabativi 's suggestion):
componentWillMount() {
    // wait for login token
    AsyncStorage.getItem('reactnativemeteor_usertoken').then((loginToken) => {
      if (loginToken) {
        // if token exists, they're logged in, take them to the initial route
        this.setState({initialRoute: Router.getRoute('drawer').routeName});
        this.setState({loggedIn: true});
      } else {
       // if token doesn't exist, they're not logged in, so take them to sign in screen
        this.setState({initialRoute: Router.getRoute('signIn').routeName});
        this.setState({loggedIn: false});
      }
      // either way, the token is done loading
      this.setState({isLoading: false});
    }).done();
  }

  render() {
    if (this.state.isLoading) return <Loading />;
    return (
      <StackNavigation initialRoute={this.state.initialRoute} />
    )


I'm using ex-navigation, so you're router may look different.

**Comment by JulianKingman** [comment]
You were returning the logged in state with a method. I'm not 100% sure, but I believe the only things that cause a react component to re-render are updates to either `state` or `props`, unless you specifically request a re-render with `shouldComponentUpdate`.

**Comment by ujwal-setlur** [comment]
I am actually following the example specified in the documentation: https://github.com/inProgress-team/react-native-meteor/blob/master/README.md#react-native-router-flux

I am starting to suspect that my Meteor.user() subscription is not populated after I log in. It's that a possibility? Is there a way to debug that to verify whether my subscription is up to date? Is there a way to force an refresh on a subscription?

**Comment by ujwal-setlur** [comment]
OK, I think the subscription is fine. Meteor.user() subscription seems to be updated just fine. So, I winder if it is an issue with the Switch component of react-native-router-flux. Has anyone used this combination successfully?

Is there a recommended router for use with react-native-meteor?

**Comment by JulianKingman** [comment]
Hmm. How is your google sign-in configured? I'm guessing it's less to do with routing than some mis-configuration. Are you saving the login token to AsyncStorage somewhere? Could it be that the sign in service is not ready until after it tries to log in?

That said, react-navigation just came out in beta, it's a pretty good router that continues the work started with ex-navigation. Check it out here: reactnavigation.org

**Comment by ujwal-setlur** [comment]
Google and Facebook signin are working just fine. After google/facebook signin, the Meteor signin is done, which is also successful. Once the Meteor signin is done, the login token is saved to AsyncStorage. It is at this point that I expect reactivity of Meteor.user() to kick in at the main component and the Switch component of the router takes the user to the home page.

Keep in mind that this works about 75% of the time. If it was always failing, I can buy into the misconfiguration theory.

I have now worked around it by forcing a redirect after I save the login token.

Here is the signin code from SignIn.js; do you see any issues with it?

```
  logUserIntoMeteor(service, user) {
    console.log("Logging " + service + " user into Meteor");
    const Data = Meteor.getData();
    if (user) {
      Meteor.call("login", {
        [service]: user
      }, (err, result) => {
        if (!err) { // save user id and token
          AsyncStorage.setItem(USER_TOKEN_KEY, result.token);
          Data._tokenIdSaved = result.token;
          Meteor._userIdSaved = result.id;
          // Force redirection to new page. Should be handled by the Switch component
          // when it detects a login, but it doesn't always work
          if (Meteor.user !== null) {
            Actions.goHome();
          }
        } else {
          console.log(err);
        }
      });
    }
  }
```

I have started looking at react-navigation. Looks interesting...

**Comment by JulianKingman** [comment]
Sounds like a good work-around. Have you tried logging the props and data from `selector` to see if the data is coming through properly?

**Comment by ujwal-setlur** [comment]
Yes, I did that, and that's why I am _reasonably_ confident that the new props are getting to the switch component. I even investigated the createContainer code of react-native-meteor, and even put in tracing code in the Switch code of react-native-router-flux, and it looks fine. I think react-native-router-flux has some issues in re-rendering.

I also looked at @Mokto's [react-native-meteor-router-flux](https://github.com/inProgress-team/react-native-meteor-router-flux). It looks very clean, but it didn't work. Not sure why, but then it is deprecated. But I was hoping it would work.

Anyway, looks like I have to live with this work-around. Bummer, I hate having to put in hacks like that :)
